Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift

```diff
@@ -0,0 +1,414 @@
+        if let name = parseOperatorName(original) {
+            print("\(name)\n\(original)\n\n")
+            possibleDisambiguationText = original[name.endIndex...]
+        } else {
+            possibleDisambiguationText = original
+        }
+        
+        // Look for the start of the parameter disambiguation.
+        if let parameterStartRange = possibleDisambiguationText.range(of: "-(") {
+            let name = original[..<parameterStartRange.lowerBound]
+            var scanner = StringScanner(original[parameterStartRange.upperBound...])
+            
+            let parameterTypes = scanner.scanArguments()
+            if scanner.isAtEnd {
+                return PathComponent(full: String(original), name: name, disambiguation: .typeSignature(parameterTypes: parameterTypes, returnTypes: nil))
+            } else if scanner.hasPrefix("->") {
+                _ = scanner.take(2)
+                let returnTypes = scanner.scanArguments() // The return types (tuple or not) can be parsed the same as the arguments
+                return PathComponent(full: String(original), name: name, disambiguation: .typeSignature(parameterTypes: parameterTypes, returnTypes: returnTypes))
+            }
+        } else if let parameterStartRange = possibleDisambiguationText.range(of: "->") {
+            let name = original[..<parameterStartRange.lowerBound]
+            var scanner = StringScanner(original[parameterStartRange.upperBound...])
+            
+            let returnTypes: [Substring]
+            if scanner.peek() == "(" {
+                _ = scanner.take() // the leading parenthesis
+                returnTypes = scanner.scanArguments() // The return types (tuple or not) can be parsed the same as the arguments
+            } else {
+                returnTypes = [scanner.takeAll()]
+            }
+            return PathComponent(full: String(original), name: name, disambiguation: .typeSignature(parameterTypes: nil, returnTypes: returnTypes))
+        }
+        
+        // This path component doesn't have type signature disambiguation.
+        return nil
+    }
+}
+
+// MARK: Scanning a substring
+
+private struct StringScanner {
+    private var remaining: Substring
+    
+    init(_ original: Substring) {
+        remaining = original
+    }
+    
+    func peek() -> Character? {
+        remaining.first
+    }
+    
+    mutating func take() -> Character {
+        remaining.removeFirst()
+    }
+    
+    mutating func take(_ count: Int) -> Substring {
+        defer { remaining = remaining.dropFirst(count) }
+        return remaining.prefix(count)
+    }
+    
+    mutating func takeAll() -> Substring {
+        defer { remaining.removeAll() }
+        return remaining
+    }
+    
+    mutating func scan(until predicate: (Character) -> Bool) -> Substring? {
+        guard let index = remaining.firstIndex(where: predicate) else {
+            return nil
+        }
+        defer { remaining = remaining[index...] }
```

Comment:
Are these two lines equivalent to calling `take(index)` ? Could you call that instead?

Reply:
Crucially, Substring (nor anything StringProtocol-y for that matter) doesn't conform to `RandomAccessCollection`, which is the protocol that makes it possible to calculate [`distance(from:to:)`](https://developer.apple.com/documentation/swift/collection/distance(from:to:)-4r6r8)) in O(1) time. The reason why is precisely because of David's explanation above.